src/topology/perfect.lean

```diff
@@ -211,3 +218,342 @@ begin
+  cases cauchy_seq_tendsto_of_complete this with y hy,
+  use y,
+  rw mem_Inter,
+  intro n,
+  apply hanti _ (x n),
+  apply mem_closure_of_tendsto hy,
+  rw [← res_succ, eventually_at_top],
+  use n.succ,
+  intros m hm,
+  exact umem _ _ hm,
+end
+
+/-- A scheme where the children of each set are pairwise disjoint induces an injective map. -/
+theorem map_injective_of_disjoint (hA : disjoint A) : injective A.map.2 :=
+begin
+  rintros ⟨x, hx⟩ ⟨y, hy⟩ hxy,
+  rw [← subtype.val_inj, ← eq_iff_res_eq],
+  intro n,
+  induction n with n ih, { simp },
+  simp only [res_succ],
+  suffices : x n = y n, { rw [ih, this] },
+  contrapose hA,
+  simp only [disjoint, ne.def, not_forall, exists_prop],
+  refine ⟨res x n, _, _, hA, _⟩,
+  rw not_disjoint_iff,
+  use A.map.2 ⟨x, hx⟩,
+  split,
+  { rw ← res_succ,
+    apply map_mem, },
+  rw [hxy, ih, ← res_succ],
+  apply map_mem,
+end
+
+end set.scheme
+end scheme
+
+section cantor_inj
+
+open function
+variables {α : Type*} [metric_space α] {C : set α} (hC : perfect C)
+include hC
+
+@[nolint ge_or_gt]
+lemma perfect.small_diam_aux (ε : ennreal) (ε_pos : ε > 0) {x : α} (xC : x ∈ C) :
+  let D := closure (emetric.ball x (ε / 2) ∩ C) in
+  perfect D ∧ D.nonempty ∧ D ⊆ C ∧ emetric.diam D ≤ ε :=
+begin
+  have : x ∈ (emetric.ball x (ε / 2)),
+  { apply emetric.mem_ball_self,
+    rw ennreal.div_pos_iff,
+    exact ⟨ne_of_gt ε_pos, by norm_num⟩, },
+  have := hC.closure_nhds_inter x xC this emetric.is_open_ball,
+  refine ⟨this.1, this.2, _, _⟩,
+  { rw is_closed.closure_subset_iff hC.closed,
+    apply inter_subset_right, },
+  rw emetric.diam_closure,
+  apply le_trans (emetric.diam_mono (inter_subset_left _ _)),
+  convert emetric.diam_ball,
+  rw [mul_comm, ennreal.div_mul_cancel]; norm_num,
+end
+
+variable (hnonempty : C.nonempty)
+include hnonempty
+
+/-- A refinement of `perfect.splitting` for metric spaces, where we also control
+the diameter of the new perfect sets. -/
+@[nolint ge_or_gt]
+lemma perfect.small_diam_splitting (ε : ennreal) (ε_pos : ε > 0) : ∃ C₀ C₁ : set α,
+  (perfect C₀ ∧ C₀.nonempty ∧ C₀ ⊆ C ∧ emetric.diam C₀ ≤ ε) ∧
+  (perfect C₁ ∧ C₁.nonempty ∧ C₁ ⊆ C ∧ emetric.diam C₁ ≤ ε) ∧ disjoint C₀ C₁ :=
+begin
+  rcases hC.splitting hnonempty with ⟨D₀, D₁, ⟨perf0, non0, sub0⟩, ⟨perf1, non1, sub1⟩, hdisj⟩,
+  cases non0 with x₀ hx₀,
+  cases non1 with x₁ hx₁,
+  rcases perf0.small_diam_aux _ ε_pos hx₀ with ⟨perf0', non0', sub0', diam0⟩,
+  rcases perf1.small_diam_aux _ ε_pos hx₁ with ⟨perf1', non1', sub1', diam1⟩,
+  refine ⟨closure (emetric.ball x₀ (ε / 2) ∩ D₀), closure (emetric.ball x₁ (ε / 2) ∩ D₁),
+    ⟨perf0', non0', sub0'.trans sub0, diam0⟩, ⟨perf1', non1', sub1'.trans sub1, diam1⟩, _⟩,
+  apply disjoint.mono _ _ hdisj; assumption,
+end
+
+open set.scheme
+
+/-- Any nonempty perfect set in a complete metric space admits a continuous injection
+from the cantor space, `ℕ → bool`. -/
+theorem exists_nat_bool_injection_of_perfect_nonempty [complete_space α]
```

Comment:
Again, leverage dot notation. I don't think it's necessary to mention the nonemptiness assumption in the name.
```suggestion
theorem perfect.exists_nat_bool_injection [complete_space α]
```